src/main/cpp/src/hive_hash.cu

```diff
@@ -486,16 +526,70 @@ std::unique_ptr<cudf::column> hive_hash(cudf::table_view const& input,
 
   check_nested_depth(input);
 
+  // `flattened_column_views` only contains nested columns and columns that result from flattening
+  // nested columns
+  std::vector<cudf::column_view> flattened_column_views;
+  // `first_child_index` has the same size as `flattened_column_views`
+  std::vector<cudf::size_type> first_child_index;
+  // `nested_column_map` has the same size as `input.num_columns()`, and it maps the column index in
+  // `input` to the index in `flattened_column_views`
+  std::vector<cudf::size_type> nested_column_map;
+
+  for (auto i = 0; i < input.num_columns(); i++) {
+    auto const& root_col = input.column(i);
+    if (root_col.type().id() == cudf::type_id::LIST ||
+        root_col.type().id() == cudf::type_id::STRUCT) {
+      // Construct the `flattened_column_views` by level order traversal
+      nested_column_map.push_back(flattened_column_views.size());
+      flattened_column_views.push_back(root_col);
+      // flattened_column_views[idx] is the next column to process
+      for (auto idx = nested_column_map.back();
+           idx < static_cast<cudf::size_type>(flattened_column_views.size());
+           idx++) {
+        auto const col = flattened_column_views[idx];
+        if (col.type().id() == cudf::type_id::LIST) {
+          first_child_index.push_back(flattened_column_views.size());
+          flattened_column_views.push_back(cudf::lists_column_view(col).offsets());
+          flattened_column_views.push_back(cudf::lists_column_view(col).get_sliced_child(stream));
+        } else if (col.type().id() == cudf::type_id::STRUCT) {
+          first_child_index.push_back(flattened_column_views.size());
+          for (auto child_idx = 0; child_idx < col.num_children(); child_idx++) {
+            flattened_column_views.push_back(
+              cudf::structs_column_view(col).get_sliced_child(child_idx, stream));
+          }
```

Comment:
We can extract `cudf::lists_column_view(col)` to be a shared variable and  `cudf::structs_column_view(col)` out of the loop.